File: src/plugin/discord/Bot.java
package plugin.discord;


import arc.util.CommandHandler;
import arc.util.Log;
import mindustry.Vars;
import mindustry.gen.Player;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;
import net.dv8tion.jda.api.requests.GatewayIntent;
import plugin.PVars;

import java.awt.*;
import java.util.EnumSet;

import static plugin.PVars.*;

public class Bot {
    public static void load() {
        EnumSet<GatewayIntent> intents = EnumSet.allOf(GatewayIntent.class);
        try {
            JDA jda = JDABuilder.create(botToken, intents)
                    .addEventListeners(new MessageListener())
                    .build();

            jda.awaitReady();

            serverGuild = jda.getGuildById(serverGuildStr);
            if(serverGuild != null) {
                serverChannel = serverGuild.getChannelById(TextChannel.class, serverChannelStr);
                logsChannel = serverGuild.getChannelById(TextChannel.class, logsChannelStr);
                votekicksChannel = serverGuild.getChannelById(TextChannel.class, votekicksChannelStr);
            } else {
                Log.err("Failed to get server guild!");
            }

            discordCommands = new CommandHandler(gamemode.botPrefix);
            Commands.register(discordCommands);

            Log.info("Bot loaded");
        } catch (Exception e) {
            Log.err("Failed to load discord bot!", e);
        }
    }

    public static void sendLog(String message) {
        if(logsChannel != null)
            logsChannel.sendMessage("["+gamemode.simpleName+"] "+message).queue();
    }

    public static void sendServerMessage(String message) {
        if(serverChannel == null) return;
        serverChannel.sendMessage(message).queue();
    }

    public static void sendJoinMessage(Player player, int id) {
        if(serverChannel == null) return;
        EmbedBuilder embed = new EmbedBuilder()
                .setColor(Color.green)
                .addField("", "[" + id + "] " + player.plainName() + " joined!", false);
        serverChannel.sendMessageEmbeds(embed.build()).queue();
    }

    public static void sendLeaveMessage(Player player, int id) {
        if(serverChannel == null) return;
        EmbedBuilder embed = new EmbedBuilder()
                .setColor(Color.red)
                .addField("", "[" + id + "] " + player.plainName() + " left!", false);
        serverChannel.sendMessageEmbeds(embed.build()).queue();
    }

    public static void reply(Message message, String content) {
        message.reply(content).queue();
    }
}

File: src/plugin/discord/Bot.kt
package plugin.discord

import arc.util.CommandHandler
import arc.util.Log
import mindustry.gen.Player
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.requests.GatewayIntent
import plugin.PVars
import plugin.discord.listeners.MessageListener
import plugin.discord.register.DiscordCommandAnnotationProcessor
import java.awt.Color
import java.util.*

const val nohornyBanButtonId = "nohornyban"
object Bot {
    fun load() {
        val intents = EnumSet.allOf(GatewayIntent::class.java)
        try {
            val jda = JDABuilder.create(PVars.botToken, intents)
                .addEventListeners(MessageListener())
                .build()

            jda.awaitReady()

            PVars.jda = jda;

            PVars.serverGuild = jda.getGuildById(PVars.serverGuildStr)
            if (PVars.serverGuild != null) {
                PVars.serverChannel =
                    PVars.serverGuild.getChannelById(TextChannel::class.java, PVars.serverChannelStr)
                PVars.logsChannel =
                    PVars.serverGuild.getChannelById(TextChannel::class.java, PVars.logsChannelStr)
                PVars.votekicksChannel =
                    PVars.serverGuild.getChannelById(TextChannel::class.java, PVars.votekicksChannelStr)
                PVars.nsfwChannel =
                    PVars.serverGuild.getChannelById(TextChannel::class.java, PVars.nsfwChannelStr)
                PVars.consoleChannel =
                    PVars.serverGuild.getChannelById(TextChannel::class.java, PVars.consoleChannelStr)
            } else {
                Log.err("Failed to get server guild!")
            }

            PVars.discordCommands = CommandHandler(PVars.gamemode.botPrefix)
            PVars.globalCommands = CommandHandler("gc!")
            DiscordCommandAnnotationProcessor.registerCommands(Commands(), PVars.discordCommands, PVars.globalCommands)

            Log.info("Registered ${PVars.discordCommands.commandList.size} discord commands")
            Log.info("Registered ${PVars.globalCommands.commandList.size} global discord commands")

            Log.info("Bot loaded")
        } catch (e: Exception) {
            Log.err("Failed to load discord bot!", e)
        }
    }

    fun sendLog(message: String?) {
        if (PVars.logsChannel != null) PVars.logsChannel.sendMessage(("[" + PVars.gamemode.simpleName + "] " + message).take(2000))
            .queue()
    }

    fun sendServerMessage(message: String) {
        if (PVars.serverChannel == null) return
        PVars.serverChannel.sendMessage(message.take(2000)).queue()
    }

    fun sendJoinMessage(player: Player, id: Int) {
        if (PVars.serverChannel == null) return
        val embed = EmbedBuilder()
            .setColor(Color.green)
            .addField("", "[" + id + "] " + player.plainName() + " joined!", false)
        PVars.serverChannel.sendMessageEmbeds(embed.build()).queue()
    }

    fun sendLeaveMessage(player: Player, id: Int) {
        if (PVars.serverChannel == null) return
        val embed = EmbedBuilder()
            .setColor(Color.red)
            .addField("", "[" + id + "] " + player.plainName() + " left!", false)
        PVars.serverChannel.sendMessageEmbeds(embed.build()).queue()
    }

    fun reply(message: Message, content: String) {
        message.reply(content.take(2000)).queue()
    }

    fun sendConsoleMessage(message: String) {
        if (PVars.consoleChannel == null) return
        PVars.consoleChannel.sendMessage(message.take(2000)).submit()
    }
}
